### PR TITLE
Document [[vm]] configuration section

### PR DESCRIPTION
**Pull request by dangra**
### Related Fly.io community and GitHub links
* https://github.com/superfly/flyctl/pull/3041
* https://github.com/superfly/flyctl/pull/3016

### Notes
n/a if none

### CODE REVIEW

**Comment by dangra** [comment]
@andie787 feel free to throw a 🍅 because of my writing style 😄. I can take it along any edits.

File: reference/configuration.html.markerb [hunk 1]
```diff
@@ -616,6 +616,73 @@ This value is only used by `fly launch` and `fly deploy` commands in case they n
 
 The value can be a string with units, like `initial_size = "30GB"` for example. The value can also be an integer, like `initial_size = 30` for example, with GB as the default unit.
 
+## The `vm` section
+
+This section defines the compute requirements required by the virtual machines used for the application.
```
Comment (andie787) [inline reference/configuration.html.markerb:621]:
```suggestion
This section defines the compute requirements for the Machines used for the application.
```

File: reference/configuration.html.markerb [hunk 2]
```diff
@@ -616,6 +616,73 @@ This value is only used by `fly launch` and `fly deploy` commands in case they n
...
+
+The default virtual machine size is `shared-cpu-1x` but it is not enforced if not specified in fly.toml. It means commands like `fly deploy` won't attempt to update the compute requirements (or size) for a machine. `fly scale` commands will use existing machine to infer new machine sizes and in the event that there are no machines, they will fallback to `shared-cpu-1x`.
```
Comment (andie787) [inline reference/configuration.html.markerb:623]:
```suggestion
The default Machine size is `shared-cpu-1x` but it is not enforced if not specified in `fly.toml`. It means commands like `fly deploy` won't attempt to update the compute requirements (or size) for a Machine. `fly scale` commands will use existing Machines to infer new Machine sizes, and in the event that there are no Machines, they will use `shared-cpu-1x`.
```

File: reference/configuration.html.markerb [hunk 3]
```diff
@@ -616,6 +616,73 @@ This value is only used by `fly launch` and `fly deploy` commands in case they n
...
+
+If you care about a predictable _scaling up from zero_ case, or if you have different compute requirements for different process groups, then it is highly recommended to add this section to fly.toml.
```
Comment (andie787) [inline reference/configuration.html.markerb:625]:
```suggestion
If you care about a predictable _scaling up from zero_ case, or if you have different compute requirements for different process groups, then it is highly recommended to add this section to `fly.toml`.
```

File: reference/configuration.html.markerb [hunk 4]
```diff
@@ -616,6 +616,73 @@ This value is only used by `fly launch` and `fly deploy` commands in case they n
...
+  gpu_kind = "a100-pcie-40gb"
+  host_dedication_id = "customer-id"
+  processes = ["app"] # optional
+```
+
+### `size`
+
+The preset size to use as base, like `size = "shared-cpu-1x"` for example.
+
+To get the full list run `fly platform vm-sizes` or check the [pricing page](/docs/about/pricing/#compute)
```
Comment (andie787) [inline reference/configuration.html.markerb:644]:
```suggestion
To get the full list, run `fly platform vm-sizes` or check the [pricing page](/docs/about/pricing/#compute).
```

File: reference/configuration.html.markerb [hunk 5]
```diff
@@ -616,6 +616,73 @@ This value is only used by `fly launch` and `fly deploy` commands in case they n
...
+
+### `memory`
+
+The memory to request. It can be expressed with units, like `memory = "2GB"` or `memory = "512mb"` for example. The value can also be an integer, like `memory = 1024` for example, with MB as the default unit.
+
+### `cpus`
+
+The number of vCPUs to request. Valid values are `1`, `2`, `4`, `8`, `16` but depends on `gpu_kind`.
```
Comment (andie787) [inline reference/configuration.html.markerb:652]:
```suggestion
The number of vCPUs to request. Valid values are `1`, `2`, `4`, `8`, or `16`, but depends on `gpu_kind`. 
```

File: reference/configuration.html.markerb [hunk 6]
```diff
@@ -616,6 +616,73 @@ This value is only used by `fly launch` and `fly deploy` commands in case they n
...
+
+The kind of CPU to request. Valid values are `shared` and `performance`.
+
+### `gpu_kind`
+
+The kind of GPU to request. Valid values are `a100-pcie-40gb` and `a100-pcie-80gb`.
+
+### `host_dedication_id`
+
+In case of having a host dedication, this is the value to set so machines are launched on dedicated hardware.
```
Comment (andie787) [inline reference/configuration.html.markerb:664]:
```suggestion
In the case of having a dedicated host, this is the value to set so machines are launched on dedicated hardware.
```

File: reference/configuration.html.markerb [hunk 7]
```diff
@@ -616,6 +616,73 @@ This value is only used by `fly launch` and `fly deploy` commands in case they n
...
+
+### `processes`
+
+Optionally, you can specify a list of [processes](#the-processes-section) to limit by process group. You can even specify two different `[[vm]]` sections to have different compute requirements per process group. Note the need to use double brackets if there are multiple `[[vm]]` defined.
```
Comment (andie787) [inline reference/configuration.html.markerb:668]:
```suggestion
Optionally, you can specify a list of [processes](#the-processes-section) to limit by process group. You can even specify two different `[[vm]]` sections to have different compute requirements per process group. Note the need to use double brackets if there are multiple `[[vm]]` sections defined.
```

File: reference/configuration.html.markerb [hunk 8]
```diff
@@ -616,6 +616,73 @@ This value is only used by `fly launch` and `fly deploy` commands in case they n
...
+
+A `[[vm]]` section without `processes` key applies to all process groups, except when another `[[vm]]` section targets a specific process group. For example the following snippets uses `shared-cpu-4x` for `app` and `other` process groups but `performance-1x` for `worker` group.
```
Comment (andie787) [inline reference/configuration.html.markerb:670]:
```suggestion
A `[[vm]]` section without a `processes` key applies to all process groups, except when another `[[vm]]` section targets a specific process group. For example the following snippets uses `shared-cpu-4x` for `app` and `other` process groups but `performance-1x` for `worker` group.
```

**Review by andie787 — APPROVED** [review]
I went super picky on you with tiny changes; but this was really good and clear.

I realize we're calling this `[vm]` to align with all our `--vm-` options in flyctl. But just want to add here that in docs we avoid saying "VM" and almost always say Machines (after we've said "Machines are our VMs" of course to make things conceptually more clear :) )

File: reference/configuration.html.markerb (hunk 3 above)
Comment (andie787) [inline reference/configuration.html.markerb:625]: I think we need to explicitly repeat here that `fly scale count` and `fly deploy` will use these settings.

Comment (dangra) [reply reference/configuration.html.markerb]: check this please https://github.com/superfly/docs/pull/1179/commits/a64821beefdf8cad71ab5896c40d67bc7c1763e9

**Pull request merged**